### PR TITLE
Adds make rule for building the rpm.

### PR DESCRIPTION
**Pull request by ilapost**
### Reason
Gatekeeper needs to be installed in core.
### Implementation
Added rules in Makefile
### Verification
Run make
### Testing
Created successfully the rpm.

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -1,8 +1,14 @@
...
+include common.mk
 
-GATEKEEPER := dist/gatekeeper
-
-all: $(GATEKEEPER)
+all: $(RPM)
 
 $(GATEKEEPER) : cmd/gatekeeper/main.go
-	$(DOCKER) gatekeeper-build go build -ldflags "-X main.BuildTime=`date -u '+%Y-%m-%d_%I:%M:%S%p'` -X main.Version=`git -C ./ describe --abbrev=0 --tags HEAD`" -a -installsuffix cgo -o src/github.com/nemosupremo/vault-gatekeeper/dist/gatekeeper github.com/nemosupremo/vault-gatekeeper/cmd/gatekeeper
+	$(DOCKER) docker.br.hmheng.io/hmheng-infra/gatekeeper-build:v1.0.3-10-ga65b145 make -C /go/src/github.com/nemosupremo/vault-gatekeeper -f Makefile.docker $@
```
Comment (kytomaki) [inline Makefile:8]: I don't think we should use internal docker repositories in a public repo.

File: Makefile [hunk 2]
```diff
@@ -1,8 +1,14 @@
-DOCKER := docker run --rm -v $(shell pwd)/../../../..:/go
+include common.mk
 
-GATEKEEPER := dist/gatekeeper
-
-all: $(GATEKEEPER)
+all: $(RPM)
```
Comment (kytomaki) [inline Makefile]: We might want to have a different `.PHONY` target for generating the rpm. e.g. `rpm`. The first (`all`) target for generating the binary could be left as is.

File: Makefile [hunk 3]
```diff
@@ -1,8 +1,14 @@
...
 
 $(GATEKEEPER) : cmd/gatekeeper/main.go
-	$(DOCKER) gatekeeper-build go build -ldflags "-X main.BuildTime=`date -u '+%Y-%m-%d_%I:%M:%S%p'` -X main.Version=`git -C ./ describe --abbrev=0 --tags HEAD`" -a -installsuffix cgo -o src/github.com/nemosupremo/vault-gatekeeper/dist/gatekeeper github.com/nemosupremo/vault-gatekeeper/cmd/gatekeeper
+	$(DOCKER) docker.br.hmheng.io/hmheng-infra/gatekeeper-build:v1.0.3-10-ga65b145 make -C /go/src/github.com/nemosupremo/vault-gatekeeper -f Makefile.docker $@
+
+$(RPM): $(GATEKEEPER)
+	$(DOCKER) docker.br.hmheng.io/hmheng-infra/fpm:latest make -C /go/src/github.com/nemosupremo/vault-gatekeeper -f Makefile.docker $@
```
Comment (kytomaki) [inline Makefile:11]: There's the internal repo again. Also we might want to simplify the build with a multistage-build?

File: Makefile.docker [hunk 4]
```diff
@@ -0,0 +1,20 @@
+include common.mk
+
+$(GATEKEEPER):
+	go get -d ./...
+	go build -ldflags "-X main.BuildTime=`date -u '+%Y-%m-%d_%I:%M:%S%p'` -X main.Version=`git -C ./ describe --abbrev=0 --tags HEAD`" -a -installsuffix cgo -o dist/gatekeeper github.com/nemosupremo/vault-gatekeeper/cmd/gatekeeper
+
+$(RPM):
```
Comment (kytomaki) [inline Makefile.docker]: `$(RPM)` target could depend on `$(GATEKEEPER)`

File: Makefile.docker [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+	mkdir -p dist/package
```
Comment (kytomaki) [inline Makefile.docker]: We could get rid of this ...

File: Makefile.docker [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+	install $(GATEKEEPER) dist/package
```
Comment (kytomaki) [inline Makefile.docker]: ... and get rid of this, if ...

File: Makefile.docker [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+	fpm \
+	  -p dist/$(RPM) \
+	  -s dir \
+	  -t rpm \
+	  --name $(RPM_NAME) \
+	  --version $(VERSION) \
+	  --iteration $(ITERATION) \
+	  --architecture x86_64 \
+	  --force \
+	  -C dist/package \
```
Comment (kytomaki) [inline Makefile.docker]: ... we get of rid of this also and ...

File: Makefile.docker [hunk 8]
```diff
@@ -0,0 +1,20 @@
...
+	  .
```
Comment (kytomaki) [inline Makefile.docker]: ... replace this with `$(GATEKEEPER)`

File: common.mk [hunk 9]
```diff
@@ -0,0 +1,6 @@
+DOCKER := docker run --rm -v $(shell pwd):/go/src/github.com/nemosupremo/vault-gatekeeper
+RPM_NAME := gatekeeper
+VERSION := 1.0.3
+ITERATION := 1
+RPM := $(RPM_NAME)-$(VERSION)-$(ITERATION).x86-64.rpm
```
Comment (kytomaki) [inline common.mk]: Architecture should be `x86_64` with an underscore

File: Makefile [hunk 10]
```diff
@@ -1,8 +1,14 @@
...
-GATEKEEPER := dist/gatekeeper
-
-all: $(GATEKEEPER)
+all: $(RPM)
 
 $(GATEKEEPER) : cmd/gatekeeper/main.go
-	$(DOCKER) gatekeeper-build go build -ldflags "-X main.BuildTime=`date -u '+%Y-%m-%d_%I:%M:%S%p'` -X main.Version=`git -C ./ describe --abbrev=0 --tags HEAD`" -a -installsuffix cgo -o src/github.com/nemosupremo/vault-gatekeeper/dist/gatekeeper github.com/nemosupremo/vault-gatekeeper/cmd/gatekeeper
+	$(DOCKER) docker.br.hmheng.io/hmheng-infra/gatekeeper-build:v1.0.3-10-ga65b145 make -C /go/src/github.com/nemosupremo/vault-gatekeeper -f Makefile.docker $@
+
+$(RPM): $(GATEKEEPER)
```
Comment (kytomaki) [inline Makefile]: I think this should be `dist/$(RPM)`

File: Makefile.docker (hunk 4 above)
Comment (kytomaki) [reply Makefile.docker]: And also be rewritten as `dist/$(RPM)`

**Review by kytomaki — CHANGES_REQUESTED** [review]
Generally looks good. Just minor nitpicking. In general I think we should get rid of bedrock dependant internal stuff on a public repo.

Also the fpm command could use maybe a `--prefix /usr/local/bin` or similar so that the gatekeeper binary wouldn't get installed on `/`

We might want to include a config file template in `/etc` provided that the gatekeeper needs one? That way the operating system will know what files belong to which package.

The commit messages could be in an imperative as they seem to be on the upstream repo.  Also some of them might be rebased maybe?

**Comment by vaine** [comment]
re: `/usr/local/bin`

That should be `/usr/bin`. That's how all other RPMs in bedrock do it.

**Comment by jaakkoo** [comment]
Do not listen to that bearded bsd -guy about the binary location. Binaries installed via package manager should be installed to `/usr/bin`

File: Makefile (hunk 1 above)
Comment (ilapost) [reply Makefile:8]: These changes are all featured in the new bedrock branch whose purpose is to build the rpm. I can try to get rid of the gatekeeper-build image, but the purpose of this branch is still for internal consumption. I am not sure if these changes are useful to the original repo and it was never my intent to publish them. I have committed changes in Gatekeeper itself straight to the master branch with the intent of maybe someday merge them but this is just for bedrock.

**Comment by kytomaki** [comment]
So at least install the binary in `/usr/bin`. The target dependancies are mostly polish. They allow make to determine if binaries are already built.

You could `install` the `$(GATEKEEPER) into `dist/packages/usr/bin` and then use the `-C` flag of fpm or use the `--prefix` to determine the same thing. Currently the rpm installs the binary into `/` which is quite unacceptable.

If you create the `dist(packages/...` hierachy we could also include `dist/package/etc/gatekeeper.conf` or similar it they exist.

**Pull request closed without merging**